### DISCUSSION TITLE
Questions about PF/IPF plotting from ODF

### DISCUSSION THREAD

**Discussion by yyydy91** [mtex-toolbox/mtex#1533] [Ask Anything]
Dear community, 

I am a beginner in texture analysis. I am using the latest version of MTEX and 2020a MatLab. The input file is a Dream3D file converted from an h5 file. I followed the example on the page and plotted the PF/IPF plots (https://mtex-toolbox.github.io/ODFPoleFigure.html). However, I am not sure if I got the right texture plots since I am not clear about the code for these calculations. The questions are about the following coding on the page:

cs = crystalSymmetry('32');
mod1 = orientation.byEuler(90*degree,40*degree,110*degree,'ZYZ',cs);
mod2 = orientation.byEuler(50*degree,30*degree,-30*degree,'ZYZ',cs);

odf = 0.2*unimodalODF(mod1) ...
  + 0.3*unimodalODF(mod2) ...
  + 0.5*fibreODF(Miller(0,0,1,cs),vector3d(1,0,0),'halfwidth',10*degree);

1. How these '90*degree,40*degree,110*degree'...are determined? Are they determined by the crystal structure? If so, is any resource I can look for the information? 
2. How the weight percentage (0.2, 0.3, 0.5) of mod1, mod2, and fibreODF were determined? I saw different numbers in other examples. Still, is any resource I can look for the information?  
3. Why only include 'Miller(0,0,1,cs)' in 'fibreODF'?

Any suggestions and comments are much appreciated! Thank you in advance!

**Comment by kilir** [comment]
Hi,
all the choices of Euler angles, weights etc are arbitrary numbers for the purpose of demonstration.
Cheers,
Rüdiger

**Reply by yyydy91** [reply]
Hi Kilir,

Thanks for the explanation. The example contains a single set of Eulerangle. How to import all the Euler into the mod1 and mod2 for ODF calculation? Thanks!

Example from document: https://mtex-toolbox.github.io/ODFPoleFigure.html

my code:

%%
file1='C:\Users\yangyang2\Desktop\Matlab LabDCT\mtex-5.6.1\mtex-5.6.1\june 22th grain feature.dream3d';
Euler= h5read(file1, '/DataContainers/DataContainer/CellData/EulerZXZ');

Euler=permute(Euler,[2 3 4 1]);

GrainID= h5read(file1, '/DataContainers/DataContainer/CellData/GrainId');

GrainID=permute(GrainID,[2 3 4 1]);

Euler_Grain = h5read(file1, '/DataContainers/DataContainer/CellFeature/EulerAngles')';
Euler_Grain(1,:)=[];

%crystal structure info of the Mg alloy
 cs = crystalSymmetry('6/mmm', [3.2093 3.2093 5.2103], 'mineral', 'Mgternary prac');

ss = specimenSymmetry('monoclinic');
ori =orientation.byEuler(Euler_Grain,cs,ss);

mod1 = orientation.byEuler(90*degree,40*degree,110*degree,Euler_Grain,cs);
mod2 = orientation.byEuler(50*degree,30*degree,-30*degree,Euler_Grain,cs);

odf = 0.2*unimodalODF(mod1)+0.3*unimodalODF(mod2)+0.5*fibreODF(Miller({1,1,-2,1},cs),vector3d.Z,'halfwidth',10*degree);


%%%%%plotPDF(odf,Miller({1,0,-1,0},{0,0,0,1},{1,1,-2,1},cs),'complete')


plotIPDF(odf,[xvector,yvector,zvector],'antipodal','silent')


setMTEXpref('defaultColorMap',LaboTeXColorMap);
mtexColorMap parula
mtexColorbar

**Reply by yyydy91** [reply]
Hi Kilir,

I tried another way using 'calcDensity' for the ODF calculation and got the result:

Euler_Grain = h5read(file1, '/DataContainers/DataContainer/CellFeature/EulerAngles')';
Euler_Grain(1,:)=[];
 cs = crystalSymmetry('6/mmm', [3.2093 3.2093 5.2103], 'mineral', 'Mgternary prac');
ori =orientation.byEuler(Euler_Grain,cs);
odf = calcDensity(ori,'exact');
plotIPDF(odf,[xvector,zvector],'antipodal')


Result:

odf = ODF (Mgternary prac → xyz)
 
  Radially symmetric portion:
    kernel: de la Vallee Poussin, halfwidth 10°
    center: Rotations: 112475 x 1
    weight: 1

**_Then I got the error message when I tried to plot the IPF:_**

Error using nfsftmex
nfsft: Input argument x must be a 2 x M double array

Error in S2FunHarmonic.quadrature (line 74)
  nfsftmex('set_x', plan, [nodes.rho'; nodes.theta']); % set vertices

Error in S2FunHarmonicSym.quadrature (line 71)
sF = S2FunHarmonic.quadrature(varargin{:});

Error in unimodalComponent/calcPDF (line 41)
  pdf = S2FunHarmonicSym.quadrature(inv(component.center)*sr,...

Error in ODF/calcPDF (line 45)
pdf = odf.weights(1) * calcPDF(odf.components{1},h,varargin{:});

Error in ODF/plotIPDF (line 43)
  p = ensureNonNeg(odf.calcPDF(h,r(i),varargin{:}));

Error in PFtrailAug18th (line 40)
plotIPDF(odf,[xvector,zvector],'antipodal')

Could you comment on the idea of fixing the error? Thank you!

**Comment by yyydy91** [comment]
Hi Kilir,

Please bear with me for one more question. I try to plot IPF through Orientations. I could get the discrete IPF (attached below) but the program crashed when I added 'contourf'. The code and error report is listed below. I am using MTEX 5.8.1 and MatLab2020a.

Code:

file1='C:\Users\yangyang2\Downloads\mtex-5.8.1\mtex-5.8.1\june 22th grain feature.dream3d';
Euler= h5read(file1, '/DataContainers/DataContainer/CellData/EulerZXZ');

Euler=permute(Euler,[2 3 4 1]);

GrainID= h5read(file1, '/DataContainers/DataContainer/CellData/GrainId');

GrainID=permute(GrainID,[2 3 4 1]);

Euler_Grain = h5read(file1, '/DataContainers/DataContainer/CellFeature/EulerAngles')';
Euler_Grain(1,:)=[];

 cs = crystalSymmetry('6/mmm', [3.2093 3.2093 5.2103], 'mineral', 'Mgternary prac');

ori =orientation.byEuler(Euler_Grain,cs);

r = vector3d.Z;
h = inv(ori) * r;
plot(h.symmetrise,'fundamentalRegion','contourf');

setMTEXpref('defaultColorMap',LaboTeXColorMap);

**_ERROR report:_**

MATLAB crash file:C:\Users\YANGYA~1\AppData\Local\Temp\matlab_crash_dump.24536-1:


--------------------------------------------------------------------------------
          Access violation detected at Tue Aug 23 11:50:50 2022 -0400
--------------------------------------------------------------------------------

Configuration:
  Crash Decoding           : Disabled - No sandbox or build area path
  Crash Mode               : continue (default)
  Default Encoding         : windows-1252
  Deployed                 : false
  Graphics Driver          : NVIDIA Corporation Quadro RTX 5000/PCIe/SSE2 Version 4.6.0 NVIDIA 512.15
  Graphics card 1          : Microsoft ( 0x0 ) Microsoft Remote Display Adapter Version 10.0.19041.1865 (2006-6-21)
  Graphics card 2          : NVIDIA ( 0x10de ) NVIDIA Quadro RTX 5000 Version 30.0.15.1215 (2022-3-17)
  JAWT Patch Status:       : Patch applied
  Java Version             : Java 1.8.0_202-b08 with Oracle Corporation Java HotSpot(TM) 64-Bit Server VM mixed mode
  MATLAB Architecture      : win64
  MATLAB Entitlement ID    : 712854
  MATLAB Root              : C:\Program Files\MATLAB\R2020a
  MATLAB Version           : 9.8.0.1417392 (R2020a) Update 4
  OpenGL                   : hardware
  Operating System         : Microsoft Windows 10 Enterprise
  Process ID               : 24536
  Processor ID             : x86 Family 6 Model 85 Stepping 7, GenuineIntel
  Session Key              : faa08726-78a0-4ef8-a364-82ca2de17df2
  Window System            : Version 10.0 (Build 19044)

Fault Count: 1


Abnormal termination:
Access violation

Register State (from fault):
  RAX = 0000000000000001  RBX = 0000000000000000
  RCX = 0000000000000000  RDX = 0000000000000000
  RSP = 0000006f427f9500  RBP = 0000000000a63000
  RSI = 0000000001f29000  RDI = 000001baa2640000
 
   R8 = 0000000000000000   R9 = 0000000000040080
  R10 = 0000000000000000  R11 = 0000006f427f9490
  R12 = 0000000000149694  R13 = 000001ba7871f7a0
  R14 = 000001ba830e4b80  R15 = 0000000000000000
 
  RIP = 000001ba29491390  EFL = 00010202
 
   CS = 0033   FS = 0053   GS = 002b

Stack Trace (from fault):
[  0] 0x000001ba29491390 C:\Users\yangyang2\Downloads\mtex-5.8.1\mtex-5.8.1\mex\S2Grid_find.mexw64+00005008 mexFunction+00000304
[  1] 0x00007ffcf9b3d406 C:\Program Files\MATLAB\R2020a\bin\win64\libmex.dll+00381958 mexfile::Inspector::needs_upgrade+00015190
[  2] 0x00007ffcf9b3e7db C:\Program Files\MATLAB\R2020a\bin\win64\libmex.dll+00387035 MexRetrieveVersion+00003563
[  3] 0x00007ffcf9b3e94c C:\Program Files\MATLAB\R2020a\bin\win64\libmex.dll+00387404 MexRetrieveVersion+00003932
[  4] 0x00007ffcf9b3eb14 C:\Program Files\MATLAB\R2020a\bin\win64\libmex.dll+00387860 MexRetrieveVersion+00004388
[  5] 0x00007ffcf9b1f819 C:\Program Files\MATLAB\R2020a\bin\win64\libmex.dll+00260121 QueryMLFcnTable_libmex+00006377
[  6] 0x00007ffcf998aee1                     bin\win64\pgo\m_dispatcher.dll+00634593 Mfh_file::dispatch_file_common+00000165
[  7] 0x00007ffcf998a571                     bin\win64\pgo\m_dispatcher.dll+00632177 Mfh_error::lookup+00006517
[  8] 0x00007ffcf998ae35                     bin\win64\pgo\m_dispatcher.dll+00634421 Mfh_file::dispatch+00000045
[  9] 0x00007ffcf2f5c6d2                            bin\win64\pgo\m_lxe.dll+00444114
[ 10] 0x00007ffcf2f5a3b4                            bin\win64\pgo\m_lxe.dll+00435124
[ 11] 0x00007ffcf304f35d                            bin\win64\pgo\m_lxe.dll+01438557 MathWorks::lxe::StartupLxeEngine+00633401
[ 12] 0x00007ffcf2f58a0b                            bin\win64\pgo\m_lxe.dll+00428555
[ 13] 0x00007ffcf2f56dcf                            bin\win64\pgo\m_lxe.dll+00421327
[ 14] 0x00007ffcf2f4dcfe                            bin\win64\pgo\m_lxe.dll+00384254
[ 15] 0x00007ffcf2f4cd7c                            bin\win64\pgo\m_lxe.dll+00380284
[ 16] 0x00007ffcf310bcf1                            bin\win64\pgo\m_lxe.dll+02211057 mwboost::archive::detail::oserializer<mwboost::archive::binaryTerm_oarchive,MathWorks::lxe::MatlabIrTree>::oserializer<mwboost::archive::binaryTerm_oarchive,MathWorks::lxe::MatlabIrTree>+00024361
[ 17] 0x00007ffcf2f4cb60                            bin\win64\pgo\m_lxe.dll+00379744
[ 18] 0x00007ffcf30aac7d                            bin\win64\pgo\m_lxe.dll+01813629 mwboost::archive::detail::oserializer<mwboost::archive::binaryTerm_oarchive,ir::IrTree>::oserializer<mwboost::archive::binaryTerm_oarchive,ir::IrTree>+00056505
[ 19] 0x00007ffcf998aee1                     bin\win64\pgo\m_dispatcher.dll+00634593 Mfh_file::dispatch_file_common+00000165
[ 20] 0x00007ffcf998a571                     bin\win64\pgo\m_dispatcher.dll+00632177 Mfh_error::lookup+00006517
[ 21] 0x00007ffcf998ae35                     bin\win64\pgo\m_dispatcher.dll+00634421 Mfh_file::dispatch+00000045
[ 22] 0x00007ffcf2f5bf68                            bin\win64\pgo\m_lxe.dll+00442216
[ 23] 0x00007ffcf2f92776                            bin\win64\pgo\m_lxe.dll+00665462
[ 24] 0x00007ffcf2f92660                            bin\win64\pgo\m_lxe.dll+00665184
[ 25] 0x00007ffcf2f5a3b4                            bin\win64\pgo\m_lxe.dll+00435124
[ 26] 0x00007ffcf30896e1                            bin\win64\pgo\m_lxe.dll+01677025 MathWorks::lxe::ShutdownLxeEngine+00076365
[ 27] 0x00007ffcf2f4c65d                            bin\win64\pgo\m_lxe.dll+00378461
[ 28] 0x00007ffcf2f4ea14                            bin\win64\pgo\m_lxe.dll+00387604
[ 29] 0x00007ffcf2f4fb2d                            bin\win64\pgo\m_lxe.dll+00391981
[ 30] 0x00007ffcf2f4f71f                            bin\win64\pgo\m_lxe.dll+00390943
[ 31] 0x00007ffcf3073390                            bin\win64\pgo\m_lxe.dll+01586064 MathWorks::lxe::StartupLxeEngine+00780908
[ 32] 0x00007ffcf2f579d2                            bin\win64\pgo\m_lxe.dll+00424402
[ 33] 0x00007ffcf2f4dcfe                            bin\win64\pgo\m_lxe.dll+00384254
[ 34] 0x00007ffcf2f4cd7c                            bin\win64\pgo\m_lxe.dll+00380284
[ 35] 0x00007ffcf2fdd12f                            bin\win64\pgo\m_lxe.dll+00971055 MathWorks::lxe::StartupLxeEngine+00165899
[ 36] 0x00007ffcf2fdd29b                            bin\win64\pgo\m_lxe.dll+00971419 MathWorks::lxe::StartupLxeEngine+00166263
[ 37] 0x00007ffcf2fdd192                            bin\win64\pgo\m_lxe.dll+00971154 MathWorks::lxe::StartupLxeEngine+00165998
[ 38] 0x00007ffcf30132f8                            bin\win64\pgo\m_lxe.dll+01192696 MathWorks::lxe::StartupLxeEngine+00387540
[ 39] 0x00007ffcf30132a6                            bin\win64\pgo\m_lxe.dll+01192614 MathWorks::lxe::StartupLxeEngine+00387458
[ 40] 0x00007ffcf4a25050                            bin\win64\pgo\m_lxe.dll+28528720 MathWorks::lxe::profiler::IMexProfilingSupport::~IMexProfilingSupport+00083312
[ 41] 0x00007ffd1627da8c C:\Program Files\MATLAB\R2020a\bin\win64\libmwbridge.dll+00187020 mnParser+00001212
[ 42] 0x00007ffcf9dc6a99   C:\Program Files\MATLAB\R2020a\bin\win64\mcr.dll+00551577 mcr_set_enableReadingFromStdin+00016745
[ 43] 0x00007ffcf9d5a05a   C:\Program Files\MATLAB\R2020a\bin\win64\mcr.dll+00106586 mcrOptions::set_use_license_manager+00071962
[ 44] 0x00007ffcf9d7b04a   C:\Program Files\MATLAB\R2020a\bin\win64\mcr.dll+00241738 mcrOptions::set_use_license_manager+00207114
[ 45] 0x00007ffcf9c358bd   C:\Program Files\MATLAB\R2020a\bin\win64\iqm.dll+00612541 iqm::PackagedTaskPlugin::execute+00000093
[ 46] 0x00007ffcf9dc084e   C:\Program Files\MATLAB\R2020a\bin\win64\mcr.dll+00526414 mcrIsExternalAvailableBuiltin+00045518
[ 47] 0x00007ffcf9c0ca03   C:\Program Files\MATLAB\R2020a\bin\win64\iqm.dll+00444931 iqm::Iqm::setupIqmFcnPtrs+00100227
[ 48] 0x00007ffcf9be4e2f   C:\Program Files\MATLAB\R2020a\bin\win64\iqm.dll+00282159 iqm::Iqm::create+00009279
[ 49] 0x00007ffcf9be4428   C:\Program Files\MATLAB\R2020a\bin\win64\iqm.dll+00279592 iqm::Iqm::create+00006712
[ 50] 0x00007ffcf9d97d06   C:\Program Files\MATLAB\R2020a\bin\win64\mcr.dll+00359686 mcrInstantiationError::operator=+00012150
[ 51] 0x00007ffcf9d9830c   C:\Program Files\MATLAB\R2020a\bin\win64\mcr.dll+00361228 mcrInstantiationError::operator=+00013692
[ 52] 0x00007ffcf9d95f0c   C:\Program Files\MATLAB\R2020a\bin\win64\mcr.dll+00352012 mcrInstantiationError::operator=+00004476
[ 53] 0x00007ffd48e0a203 C:\Program Files\MATLAB\R2020a\bin\win64\mwboost_thread-vc141-mt-x64-1_70.dll+00041475 mwboost::thread::swap+00000083
[ 54] 0x00007ffd5b0f1bb2                   C:\WINDOWS\System32\ucrtbase.dll+00138162 configthreadlocale+00000146
[ 55] 0x00007ffd5b507034                   C:\WINDOWS\System32\KERNEL32.DLL+00094260 BaseThreadInitThunk+00000020
[ 56] 0x00007ffd5d2c2651                      C:\WINDOWS\SYSTEM32\ntdll.dll+00337489 RtlUserThreadStart+00000033


This error was detected while a MEX-file was running. If the MEX-file
is not an official MathWorks function, please examine its source code
for errors. Please consult the External Interfaces Guide for information
on debugging MEX-files.


![Discrete IPF](https://user-images.githubusercontent.com/108360874/186204652-77384fa7-19a5-424f-b1bc-ed0d9b95d0ec.png)
